src/tokio/caller.rs

```diff
@@ -0,0 +1,135 @@
+use crate::packet::PacketCodec;
+use crate::protocol::sender::{Sender, SenderAlgorithmAction};
+use crate::protocol::Handshake;
+use crate::{SrtCongestCtrl, ControlPacket, DataPacket, Packet};
+use bytes::Bytes;
+use failure::Error;
+use futures::channel::{mpsc, oneshot};
+use futures::future::Fuse;
+use futures::select;
+use futures::{FutureExt, SinkExt, StreamExt};
+use log::debug;
+use std::net::SocketAddr;
+use std::time::{Duration, Instant};
+use tokio::net::UdpSocket;
+use tokio::time::delay_until;
+use tokio_util::udp::UdpFramed;
+
+pub struct SrtCaller {
+    close_tx: Option<oneshot::Sender<()>>,
+    srt_send_tx: mpsc::Sender<(Bytes, Instant)>,
+    _task: tokio::task::JoinHandle<()>,
+}
+
+impl SrtCaller {
+    pub async fn connect(remote: SocketAddr) -> Result<SrtCaller, Error> {
+        // TODO: unify errors and remove panics
+
+        let local = "0.0.0.0:0".parse::<SocketAddr>().unwrap();
+        let latency = Duration::from_millis(50);
+
+        let mut socket = UdpFramed::new(UdpSocket::bind(local).await.unwrap(), PacketCodec {});
+        let connect = crate::pending_connection::connect(
+            &mut socket,
+            remote,
+            rand::random(),
+            local.ip(),
+            latency,
+            None,
+        )
+        .await
+        .unwrap();
+
+        let (close_tx, close_rx) = oneshot::channel();
+        let (srt_send_tx, srt_send_rx) = mpsc::channel::<(Bytes, Instant)>(10);
+
+        let sender = Sender::new(
+            connect.settings,
+            // TODO: use handshake method from sender
+            Handshake::Caller,
+            SrtCongestCtrl,
+        );
+
+        let mut state = SrtConnectionState {
+            // this needs to be fused here so it actually doesn't get polled after
+            // completion
+            // the streams don't need this as we construct a new futures each time[
+            close_rx: close_rx.fuse(),
+            srt_send_rx,
+            socket,
+            sender,
+        };
+
+        let task = tokio::spawn(async move {
+            state.run().await;
+        });
+
+        Ok(SrtCaller {
+            close_tx: Some(close_tx),
+            srt_send_tx: srt_send_tx,
+            _task: task,
+        })
+    }
+
+    pub async fn send(&mut self, data: Bytes) -> Result<(), mpsc::SendError> {
+        self.srt_send_tx.send((data, Instant::now())).await
+    }
+
+    pub async fn shutdown(&mut self) -> Result<(), Error> {
+        self.close_tx = None;
```

Comment:
A comment here saying this is droping the `close_rx` would be good, it's not that explicit